feat: Add metadata fields to core entities

novakzaballa opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes
FE changes:
- Add and create metadata to Environments, segments, and Flags
- The feature was wrapped inside the 'enable_metadata' flag
BE changes:
- Support metadata for segments, and Flags
- Support required metadata model fields for segments, and Flags

## How did you test this code?

- Go to manage
- Click the Metadata tab.
- Create Metadata.
- Edit metadata and enable any entity

matthewelwell commented:
@novakzaballa the main thing I have found so far is that there is no feedback when you create a new metadata item. The side panel doesn't close and there is no reloading of the list. It does also seem that perhaps it's not always reading / writing to the same organisation but I'm not able to consistently reproduce that. 

https://github.com/Flagsmith/flagsmith/assets/14089968/08e16122-ae02-4ad3-a6a4-1e430a0fbfd9

novakzaballa commented:
> @novakzaballa the main thing I have found so far is that there is no feedback when you create a new metadata item. The side panel doesn't close and there is no reloading of the list. 
> 
This has already been corrected.

> It does also seem that perhaps it's not always reading / writing to the same organisation but I'm not able to consistently reproduce that.
>

I'll take a look to see if I can reproduce what you mentioned.

matthewelwell commented on frontend/web/components/modals/CreateMetadata.tsx:
```diff
@@ -0,0 +1,316 @@
+        }}
+        type='text'
+        id='metadata-name'
+        placeholder='FL-124'
```
```suggestion
        placeholder='e.g. JIRA Ticket Number'
```

matthewelwell commented on frontend/web/components/modals/CreateMetadata.tsx:
```diff
@@ -0,0 +1,316 @@
+        }}
+        type='text'
+        title={'Description (optional)'}
+        placeholder='This is a metadata description'
```
```suggestion
        placeholder='e.g. \'The JIRA Ticket Number associated with this flag\''
```

novakzaballa replied:
Done.

matthewelwell reviewed (changes requested):
Some more issues I've found in manual testing: 

<img width="1170" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/0b4c95d0-ff08-48ca-b299-07493dac5640">

This UI is kind of odd, especially since these toggles aren't even useable. Can we just use some symbols here to determine if it's enabled / required and add an edit option at the end of the row, next to 'Remove' ? 

https://github.com/Flagsmith/flagsmith/assets/14089968/4f5188af-68cb-4744-ac38-4c8c104af3ce

I can't seem to make a metadata field required for any entity. 

<img width="661" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/d41de2ff-8828-4ce9-8cf5-50cb1613b915">

The UI lets me configure the metadata as required (although it doesn't persist it, see above) even if the field is not enabled for that entity type.

novakzaballa commented on frontend/web/components/modals/CreateFlag.js:
```diff
@@ -460,6 +477,8 @@ const CreateFlag = class extends Component {
     } = this.state
     const FEATURE_ID_MAXLENGTH = Constants.forms.maxLength.FEATURE_ID
 
+    // const metadataList = this.getApprovals(metadata, this.state.metadata || [])
```
Removed.

matthewelwell reviewed (changes requested):
I haven't been through the rest of the PR, but just finishing my review here as we need to decide on whether we should include the addition of Feature and Segment metadata as part of this PR or another PR. 

I have a feeling that once you update the tests that I have added comments to, there will be a substantial amount of additional work (primarily serializers I think) to get that functionality working which may blow out this PR.

matthewelwell reviewed (changes requested):
Added some comments on the code below, but also some general feedback from functional testing. 

1. I think this would look nicer without the 'Enabled' / 'Disabled' / 'Edit Metadata' text? 

<img width="1165" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/302e82b9-dfd9-44ec-acdb-194d9ab614f9">

2. I still can't persist a required metadata field. I toggle the switcher to make it 'required' for environment, hit save then open up the metadata again and it's not required. 
3. We should add a confirm step to deleting a metadata item. 
4. I'm not sure why we have a '*' here: 

<img width="170" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/2c3a97d9-40e0-4033-84d9-958c35867f16">

5. I can't add metadata to an entity. Clicking on the metadata item doesn't do anything. This is the case for both features and environments. 

https://github.com/Flagsmith/flagsmith/assets/14089968/1e2b61ee-2b27-4e47-be63-865c004994ff

6. The search input doesn't seem to be hooked up to anything: 

<img width="526" alt="image" src="https://github.com/Flagsmith/flagsmith/assets/14089968/95dca539-cde6-4b34-977e-2bf4a1132348">

7. I understood that the Metadata items were supposed to live at the organisation level, not the project. They can be _required_ for a given entity type for a project, but they should live at the organisation level. Correct me if I'm wrong here @gagantrivedi.

dabeeeenster commented:
Should we make metadata definitions not-updateable/write once? What happens if you try and change a `string` for an `int` etc? It just feels dangerous/not applicable?

novakzaballa commented:
> Added some comments on the code below, but also some general feedback from functional testing.
>
> 1. I think this would look nicer without the 'Enabled' / 'Disabled' / 'Edit Metadata' text?

Yes, texts were deleted.


> 3. We should add a confirm step to deleting a metadata item.

Confirm modal added.

> 4. I'm not sure why we have a '*' here:

Deleted


> 6. The search input doesn't seem to be hooked up to anything:

Corrected

The other 2 items are still in progress

novakzaballa commented on frontend/common/types/responses.ts:
```diff
@@ -376,6 +376,8 @@ export type Res = {
   getSubscriptionMetadata: { id: string }
   environment: Environment
+  metadataModelField: { id: string }
+  metadata: { id: string }
```
Right, Done.

kyle-ssg commented on frontend/common/constants.ts:
```diff
@@ -212,6 +212,12 @@ export default {
     white: '#ffffff',
   },
+  contentTypes: {
+    environment: 30,
```
Are these hardcoded ids ? Not clear what this is. If they are ids, does this differ per environment or are they somehow fixed.

novakzaballa replied:
These are the IDs of the content types of those 4 entities, they do not change.
<img width="328" alt="Screen Shot 2023-12-14 at 16 32 38" src="https://github.com/Flagsmith/flagsmith/assets/41410593/d04125dc-6364-42e7-9960-35eb320173e1">

kyle-ssg commented on frontend/common/constants.ts:
```diff
@@ -466,6 +472,7 @@ export default {
       'Set different values for your feature based on what segments users are in.',
     TAGS_DESCRIPTION:
       'Organise your flags with tags, tagging your features as "<strong>protected</strong>" will prevent them from accidentally being deleted.',
+    TOOLTIP_METADATA_DESCRIPTION: 'Add metadata in your',
```
Doesn't look complete

novakzaballa replied:
It is not incomplete, it is used like this `tooltip={${Constants.strings.TOOLTIP_METADATA_DESCRIPTION} segments}`

kyle-ssg commented:
> <img alt="image" width="1165" src="https://private-user-images.githubusercontent.com/14089968/283875104-302e82b9-dfd9-44ec-acdb-194d9ab614f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3ODc5NTUsIm5iZiI6MTcwMTc4NzY1NSwicGF0aCI6Ii8xNDA4OTk2OC8yODM4NzUxMDQtMzAyZTgyYjktZGZkOS00NGVjLWFjZGItMTk0ZDlhYjYxNGY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDE0NDczNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyMjY4NjRiNGI3ZTRmZGQ2YjNhYzIwNmNkNGQwYTU1MWNkN2MyYjJlMzA0ZmUxNjJkM2ZjMTA3NzI5OGFhY2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.23F1AJxjPvebdb9yitZBZXu8i4kEZz_eWE33j2RLi0k">

I think this needs to go to design or at least more closely use the design system, e.g. making these column items checkboxes rather than the current ui.

Do we explain what "Metadata" is where the user can create it ? Maybe a subtitle and a link to docs would be useful here.

novakzaballa commented:
> 
> 
> I think this needs to go to design or at least more closely use the design system, e.g. making these column items checkboxes rather than the current ui and a less wordy edit cta.
> 

I made some changes.
> Do we explain what "Metadata" is where the user can create it ? Maybe a subtitle and a link to docs would be useful here.

I put a metadata subtitle, in the sections "managing features", "managing segments" and "environment settings"  inside the docs.

kyle-ssg commented on frontend/web/components/MetadataSelect.tsx:
```diff
@@ -0,0 +1,81 @@
+      title='Metadata'
+      isOpen={isOpen}
+      onClose={onToggle}
+      className='inline-modal--tags'
```
Can this be replaced with inline-modal--sm

novakzaballa replied:
Done

kyle-ssg commented on frontend/web/components/MetadataSelect.tsx:
```diff
@@ -0,0 +1,81 @@
+        className='full-width mb-2'
+        placeholder='Type or choose a Metadata'
+      />
+      <div style={{ maxHeight: 200, overflowY: 'auto' }}>
```
Can the inline style be replaced with className="inline-modal__list"

novakzaballa replied:
Done

kyle-ssg commented on frontend/web/components/MetadataSelect.tsx:
```diff
@@ -0,0 +1,81 @@
+                space
+              >
+                <Flex
+                  className={value?.includes(v.id) ? 'font-weight-bold' : ''}
```
If possible can we rename and reuse TableFilterItem.tsx? It does the exact same as this.

kyle-ssg commented on frontend/web/components/MyMetadataSelect.tsx:
```diff
@@ -0,0 +1,37 @@
+import { FC, useEffect, useState } from 'react'
```
Is there a reason why it's called MyMetadata, that would indicate to me it's specific to the user but I think this is for the org?

kyle-ssg commented on frontend/web/components/modals/CreateFlag.js:
```diff
@@ -531,6 +576,73 @@ const CreateFlag = class extends Component {
+                              <span className='chip-icon ion'>
+                                <IonIcon
+                                  icon={closeIcon}
+                                  style={{ fontSize: '13px' }}
```
className="fs-caption"

novakzaballa replied:
Done

kyle-ssg commented on frontend/web/components/pages/ProjectSettingsPage.js:
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
+                                      width: '185px',
+                                    }}
+                                  >
+                                    <div className='table-column'>
```
Markup here doesn't lookl correct, we are nesting table-column elements. Please lets use classnames where possible too.

```
                                    <div className='table-column d-flex text-center'>

```

novakzaballa replied:
Done

kyle-ssg commented on frontend/web/components/pages/ProjectSettingsPage.js:
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
+                                    }}
+                                  >
+                                    <div className='table-column'>
+                                      {metadata.content_type_fields.find(
```
Hard to read what the following functions are doing, perhaps separate these into variables above the markup ?

kyle-ssg commented on frontend/web/components/pages/ProjectSettingsPage.js:
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
+                                      className='table-column'
+                                      style={{ width: '150px' }}
+                                    >
+                                      {metadata.content_type_fields.find(
```
Same here

kyle-ssg commented on frontend/web/components/pages/ProjectSettingsPage.js:
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
+                                      }}
+                                    >
+                                      {metadata.content_type_fields.find(
+                                        (m) =>
```
And here

novakzaballa commented on frontend/web/components/MetadataSelect.tsx:
```diff
@@ -0,0 +1,81 @@
+        placeholder='Type or choose a Metadata'
+      />
+      <div style={{ maxHeight: 200, overflowY: 'auto' }}>
+        {metadataList &&
```
Done
